core/services/ocr2/plugins/generic/oraclefactory.go

```diff
@@ -0,0 +1,155 @@
+	"github.com/smartcontractkit/chainlink/v2/core/services/telemetry"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/core"
+)
+
+type OracleIdentity struct {
+	EVMKey                    string   `json:"evm_key"`
+	PeerID                    string   `json:"peer_id"`
+	PublicKey                 []byte   `json:"public_key"`
+	OffchainPublicKey         [32]byte `json:"offchain_public_key"`
+	ConfigEncryptionPublicKey [32]byte `json:"config_encryption_public_key"`
+}
+
+type oracleFactory struct {
+	database    ocr3types.Database
+	jobID       int32
+	jobName     string
+	jobORM      job.ORM
+	kb          ocr2key.KeyBundle
+	lggr        logger.Logger
+	config      job.OracleFactoryConfig
+	peerWrapper *ocrcommon.SingletonPeerWrapper
+	relayerSet  *RelayerSet
+	identity    OracleIdentity
+}
+
+type OracleFactoryParams struct {
+	JobID       int32
+	JobName     string
+	JobORM      job.ORM
+	Kb          ocr2key.KeyBundle
+	Logger      logger.Logger
+	Config      job.OracleFactoryConfig
+	PeerWrapper *ocrcommon.SingletonPeerWrapper
+	RelayerSet  *RelayerSet
+	Identity    OracleIdentity
+}
+
+func NewOracleFactory(params OracleFactoryParams) (core.OracleFactory, error) {
+	return &oracleFactory{
+		database:    OracleFactoryDB(params.JobID, params.Logger),
+		jobID:       params.JobID,
+		jobName:     params.JobName,
+		jobORM:      params.JobORM,
+		kb:          params.Kb,
+		lggr:        params.Logger,
+		config:      params.Config,
+		peerWrapper: params.PeerWrapper,
+		relayerSet:  params.RelayerSet,
+		identity:    params.Identity,
+	}, nil
+}
+
+type JSONConfig map[string]interface{}
+
+// Bytes returns the raw bytes
+func (r JSONConfig) Bytes() []byte {
+	b, _ := json.Marshal(r)
+	return b
+}
+
+func (of *oracleFactory) NewOracle(ctx context.Context, args core.OracleArgs) (core.Oracle, error) {
+	if !of.peerWrapper.IsStarted() {
+		return nil, errors.New("peer wrapper not started")
+	}
+
+	relayer, err := of.relayerSet.Get(ctx, types.RelayID{Network: of.config.Network, ChainID: of.config.ChainID})
+	if err != nil {
+		return nil, fmt.Errorf("error when getting relayer: %w", err)
+	}
+
+	var relayConfig = struct {
+		ChainID                string   `json:"chainID"`
+		EffectiveTransmitterID string   `json:"effectiveTransmitterID"`
+		SendingKeys            []string `json:"sendingKeys"`
+	}{
+		ChainID:                of.config.ChainID,
+		EffectiveTransmitterID: of.identity.EVMKey,
+		SendingKeys:            []string{of.identity.EVMKey},
+	}
+	relayConfigBytes, err := json.Marshal(relayConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error when marshalling relay config: %w", err)
+	}
+
+	pluginProvider, err := relayer.NewPluginProvider(ctx, core.RelayArgs{
+		ContractID:   of.config.OCRContractAddress,
+		ProviderType: "plugin",
+		RelayConfig:  relayConfigBytes,
+	}, core.PluginArgs{
+		TransmitterID: of.identity.EVMKey,
+	})
```

Comment:
Lots of chain specific details leaking here: can we somehow rewrite this so we don't expose those details in the core node?

See the example for how de instantiate the plugin provider for newServicesGenericPlugin in ocr2/delegate.go

for some fields, eg. transmitterID, it's as simple as changing the naming I think.

Reply:
https://github.com/smartcontractkit/chainlink/blob/002296d9dbbcb356d6f66d5e274b18bfb16681be/core/services/ocr2/delegate.go#L633